teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,152 @@
+
+## Proposal
+We propose to support `retries` in TaskRuns, and solidify that `Timeout` is for each Retry attempt in CustomRun by updating [TEP-0069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) and the Developer guidance [docs/runs.md](https://github.com/tektoncd/pipeline/blob/0dd9f9719177054bdf60928ce9b4aab4b85af5bd/docs/runs.md). 
+
+### Notes and Caveats
+One caveat is that **it's hard for PipelineRun Controller to know whether retries is implemented correctly by customers**. If customers want to use a specific CustomRun, they need to check (the documentation or even the code) whether the functionality is supported properly. 
+
+An [alternative](#alternatives) is removing support for retries from TaskRun and CustomRuns, make it a concern of PipelineRun.
+
+Another [alternative](#alternatives) is making retries a specialized field in each CustomRun individually instead of a common field for all CustomRuns.
+
+Another caveat is updating `retriesStatus` can lead to Large CRD issue, which can be alleviated by reducing the information we stored in `retriesStatus`, like removing taskSpec etc. We can design on it (similar to [TEP-0100](https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md)) when it becomes an actual issue.
+
+## Design Details
+
+### API Change
+#### New `retries` field in TaskRun.Spec
+
+```golang
+// TaskRunSpec defines the desired state of TaskRun
+type TaskRunSpec struct {
+  // ...
+  Retries int `json:"retries,omitempty"`
+  // ...
+}
+```
+
+Implements Retries in TaskRun reconciler, use `retriesStatus` to decide whether the retries has been exhausted.
+
+We pass [PipelineTask.Retries](https://github.com/tektoncd/pipeline/blob/481def234d1e8715ea5f0299713592f311340286/pkg/apis/pipeline/v1beta1/pipeline_types.go#L187) to TaskRunSpec.Retries and CustomRunSpec.Retries to retry TaskRuns and CustomRuns belonging to a parent PipelineRun.
+
+## Alternatives
+
+### Support retries for TaskRun and CustomRun in PipelineRun Controller
+
+Remove support for retries in TaskRun and CustomRun, make retries a PipelineRun concern; move logic for retries to PipelineRun reconciler and create new Runs at each attempt (remove the RetriesStatus); Pipeline Task Timeout applies to each retry individually.
+
+In this way, we don't need to worry about whether the retries is implemented correctly in CustomRun.
+
+However, TaskRun and CustomRun **can't retry on their own**, as Runtime Objects that can run on their own, we expect they can also be retried to avoid transient errors. There are use cases implying that we need it:
+  - Run: [PipelineLoop](https://github.com/tektoncd/experimental/blob/main/pipeline-loops/examples/pipelinespec-with-run-arrary-value.yaml)
+  - TaskRuns
```

Comment:
I'm still not completely convinced that there's a use case for taskruns to retry on their own-- nobody has requested this feature AFAIK and I'd imagine that most taskruns people want to retry are part of a pipelinerun. However it does seem reasonable that some people might want to retry the taskruns listed here.

Reply:
Let's move the conversation to [this comment thread](https://github.com/tektoncd/community/pull/816#discussion_r976907440) for better discussion.